src/read_input_stream.rs

```diff
@@ -0,0 +1,411 @@
+                    };
+
+                    loop {
+                        match std_error_to_gio_error(read.seek(pos)) {
+                            None => continue,
+                            Some(res) => return res.map(|_| ()),
+                        }
+                    }
+                }
+                _ => Err(glib::Error::new(
+                    crate::IOErrorEnum::NotSupported,
+                    "Truncating not supported",
+                )),
+            }
+        }
+
+        fn can_truncate(&self, _seekable: &crate::Seekable) -> bool {
+            false
+        }
+
+        fn truncate(
+            &self,
+            _seekable: &crate::Seekable,
+            _offset: i64,
+            _cancellable: Option<&crate::Cancellable>,
+        ) -> Result<(), glib::Error> {
+            Err(glib::Error::new(
+                crate::IOErrorEnum::NotSupported,
+                "Truncating not supported",
+            ))
+        }
+    }
+}
+
+glib_wrapper! {
+    pub struct ReadInputStream(Object<subclass::simple::InstanceStruct<imp::ReadInputStream>, subclass::simple::ClassStruct<imp::ReadInputStream>, ReadInputStreamClass>) @extends crate::InputStream, @implements crate::Seekable;
+
+    match fn {
+        get_type => || imp::ReadInputStream::get_type().to_glib(),
+    }
+}
+
+impl ReadInputStream {
+    pub fn new<R: Read + Send + 'static>(read: R) -> ReadInputStream {
+        let obj = glib::Object::new(Self::static_type(), &[])
+            .expect("Failed to create read input stream")
+            .downcast()
+            .expect("Created read input stream is of wrong type");
+
+        let imp = imp::ReadInputStream::from_instance(&obj);
+        *imp.read.borrow_mut() = Some(imp::Reader::Read(AnyReader::new(read)));
+
+        obj
+    }
+
+    pub fn new_seekable<R: Read + Seek + Send + 'static>(read: R) -> ReadInputStream {
+        let obj = glib::Object::new(Self::static_type(), &[])
+            .expect("Failed to create read input stream")
+            .downcast()
+            .expect("Created read input stream is of wrong type");
+
+        let imp = imp::ReadInputStream::from_instance(&obj);
+        *imp.read.borrow_mut() = Some(imp::Reader::ReadSeek(AnyReader::new_seekable(read)));
+
+        obj
+    }
+
+    pub fn close_and_take(&self) -> Box<dyn Any + Send + 'static> {
+        let imp = imp::ReadInputStream::from_instance(self);
+        let inner = imp.read.borrow_mut().take();
+
+        let ret = match inner {
+            None => {
+                panic!("Stream already closed or inner taken");
+            }
+            Some(imp::Reader::Read(read)) => read.reader,
+            Some(imp::Reader::ReadSeek(read)) => read.reader,
+        };
+
+        let _ = self.close(crate::NONE_CANCELLABLE);
+
+        match ret {
+            AnyOrPanic::Any(r) => r,
+            AnyOrPanic::Panic(p) => std::panic::resume_unwind(p),
+        }
+    }
+}
+
+enum AnyOrPanic {
```

Comment:
Duplicate?

Reply:
You mean the two versions in `read_input_stream.rs` and `write_output_stream.rs`? Yes, it seemed cleaner to duplicate those 4 lines than exporting them from the module. At later times we might also need to have them look different.

Reply:
Ok, it private anyway